src/simulate.py

```diff
@@ -93,12 +93,27 @@ def simulate_cycles(self, hw_spec, computation_graph, total_cycles):
                 #     print(
                 #         f" found a mem object in active energy calc, adding scaling factor: {scaling}"
                 #     )
-
-            self.total_energy += (
-                hw_spec.dynamic_power[node_data["function"]] * 1e-9
-                * scaling
-                * (hw_spec.latency[node_data["function"]] / hw_spec.frequency)
-            )
+            # ADDED
+            if node_data["function"] in ["Buf", "MainMem"]:
+                self.total_energy += (
+                    (hw_spec.dynamic_energy[node_data["function"]]["Read"] 
+                    + hw_spec.dynamic_energy[node_data["function"]]["Write"])
+                    * 1e-9
+                    * scaling
+                )
+                if node_data["function"] == "MainMem":
+                    self.total_energy += (
+                    hw_spec.dynamic_power["OffChipIO"] * 1e-9
+                    * scaling
+                    * (hw_spec.latency["OffChipIO"] / hw_spec.frequency)
```

Comment:
the `IOTime` you return is in units of seconds. so you don't need to divide again by frequency here.